flink-table/flink-table-common/src/main/java/org/apache/flink/table/functions/LegacyUserDefinedFunctionInference.java

```diff
@@ -0,0 +1,529 @@
+        return new InputTypeStrategy() {
+            @Override
+            public ArgumentCount getArgumentCount() {
+                return ConstantArgumentCount.any();
+            }
+
+            @Override
+            public Optional<List<DataType>> inferInputTypes(
+                    CallContext callContext, boolean throwOnFailure) {
+                final List<DataType> argumentDataTypes = callContext.getArgumentDataTypes();
+                final LogicalType[] input =
+                        argumentDataTypes.stream()
+                                .map(DataType::getLogicalType)
+                                .toArray(LogicalType[]::new);
+
+                final Optional<Method> foundMethod =
+                        getUserDefinedMethod(
+                                func,
+                                "eval",
+                                logicalTypesToExternalClasses(input),
+                                input,
+                                (paraClasses) ->
+                                        Arrays.stream(func.getParameterTypes(paraClasses))
+                                                .map(TypeConversions::fromLegacyInfoToDataType)
+                                                .toArray(DataType[]::new));
+
+                if (foundMethod.isPresent()) {
+                    return Optional.of(argumentDataTypes);
+                } else {
+                    return callContext.fail(
+                            throwOnFailure, "Given parameters do not match any signature.");
+                }
+            }
+
+            @Override
+            public List<Signature> getExpectedSignatures(FunctionDefinition definition) {
+                return getSignatures(func, "eval");
+            }
+        };
+    }
+
+    private static Optional<Class<?>[]> getEvalMethodSignature(
+            ScalarFunction func, LogicalType[] expectedTypes) {
+        return getUserDefinedMethod(
+                        func,
+                        "eval",
+                        logicalTypesToExternalClasses(expectedTypes),
+                        expectedTypes,
+                        (paraClasses) ->
+                                Arrays.stream(func.getParameterTypes(paraClasses))
+                                        .map(TypeConversions::fromLegacyInfoToDataType)
+                                        .toArray(DataType[]::new))
+                .map(
+                        m ->
+                                getParamClassesConsiderVarArgs(
+                                        m.isVarArgs(),
+                                        m.getParameterTypes(),
+                                        expectedTypes.length));
+    }
+
+    private static Class<?>[] getParamClassesConsiderVarArgs(
+            boolean isVarArgs, Class<?>[] matchingSignature, int expectedLength) {
+        return IntStream.range(0, expectedLength)
```

Comment:
(nit) No need to create a stream to iterate and then make an array.  

Looks like https://docs.oracle.com/javase/8/docs/api/java/util/Arrays.html#setAll-T:A-java.util.function.IntFunction- may be an option for doing this in one step / re-using the lambda.

(I don't necessarily think this needs to be addressed.)

Reply:
Not sure if that's a better approach. 

To be honest I like the single chain approach a bit more. What you're suggesting would look:
```
final Class<?>[] array = new Class<?>[types.length];
Arrays.setAll(array, i -> {
        final LogicalType t = types[i];
        if (t == null) {
            return null;
        } else {
            return TypeConversions.fromLogicalToDataType(t)
                    .getConversionClass();
        }
    });
return array;
```

Reply:
Here it won't matter (much).

In the old days, Scala could generate tons of intermediate objects to handle things like this.  That got me used to looking for when objects were being created just to perform iteration.  

This code just reminded me of situations like that!